src/serialization/sb3.js

```diff
@@ -974,18 +1248,183 @@ const parseScratchObject = function (object, runtime, extensions, zip, assets) {
         // Take a second pass to create objects and add extensions
         for (const blockId in object.blocks) {
             if (!Object.prototype.hasOwnProperty.call(object.blocks, blockId)) continue;
             const blockJSON = object.blocks[blockId];
-            blocks.createBlock(blockJSON);
+            let version = 0;
+            const blockOpcode = blockJSON.opcode;
+
+            // Check if version name is included
+            const regex = /_v(\d+)/g;
+            const matches = blockOpcode.match(regex); // Get all matches
+
+            if (matches) {
+                const lastMatch = matches[matches.length - 1]; 
+                const versionMatch = lastMatch.match(/_v(\d+)/); 
+
+                if (versionMatch) {
+                    version = parseInt(versionMatch[1], 10); // Extract and parse the version number
+                }
+                blockJSON.opcode = blockOpcode.replace(regex, ""); // Remove all version numbers from the opcode
+            }
 
-            // If the block is from an extension, record it.
             const extensionID = getExtensionIdForOpcode(blockJSON.opcode);
+            const menuRegex = /menu_\d+/g;
+            
+
+            // Make sure the block is one that can't be versioned
+            if (extensionID && !Object.keys(primitiveOpcodeInfoMap).includes(blockJSON.opcode) && !menuRegex.test(blockJSON.opcode)) {
+                // Collect block version information for extension
+                const instance = extensionManager.getExtensionInstance(extensionID);
+                
+                var extensionBlocks = instance.info.blocks;
+                extensionBlocks = extensionBlocks.reduce((acc, tempBlock) => {
+                    acc[tempBlock.opcode] = tempBlock;
+                    return acc;
+                }, {});
+                var blockInfoIndex = blockJSON.opcode;
+                blockInfoIndex = blockJSON.opcode.replace(`${blockJSON.opcode.split("_")[0]}_`, "");
+                const versionMap = createNameMap(extensionBlocks);
+                if (versionMap[version] && versionMap[version][blockInfoIndex]) {
+                    blockInfoIndex = versionMap[version][blockInfoIndex];
+                }
+                const versionList = extensionBlocks[blockInfoIndex].versions;
+                // Make sure version functions exist
+                if (versionList != [] && version < Object.keys(versionList).length) {
+                    // Delete static images
+                    const blockArgs = removeImageEntries(extensionBlocks[blockInfoIndex].arguments);
+                    // Gather arguments and variable positions
+                    var { inputs, variables } = gatherInputs(object.blocks, blockJSON);
+                    var fields = gatherFields(blockJSON, blockArgs);
+                    var totalList = addInputsAndFields(inputs, fields, blockArgs);
+                    const newInputs = {};
```

Comment:
FYI Not sure about this specifically, but for a lot of this storage objects, it'd lead to easier to read code to use a [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) instead.

And if you later need an object, you can always go: `const obj = Object.fromEntries(map);`